### DISCUSSION TITLE
Responsive navigator

### DISCUSSION THREAD

**Discussion by cglacet** [react-navigation/react-navigation#11661] [Q&A]
I have an app that has a two kinds of navigator depending on the device its running on, something like this:

```tsx
function RootNavigator() {
    const {isDesktop} = useDeviceQuery();
    return <Stack.Navigator screenOptions={{headerShown: false}} initialRouteName="Root">
        <Stack.Screen
            name="Root"
            component={isDesktop ? BottomTabNavigator : DrawerNavigator}
        />
    </Stack.Navigator>
}
```

This root navigator is then simply nested under the `NavigationContainer`.

The problem is when switching from desktop to mobile navigation, the whole application reloads and we lost track of the navigation. This can for example happen when the user is on desktop and modify its browser window size, the windows could get to the size where its considered too small to render the desktop navigation and start showing the mobile (drawer) navigation instead.

Someone with a similar issue on stackoverflow: [Responsive Navigator in React Navigation](https://stackoverflow.com/questions/66026066/responsive-navigator-in-react-navigation)

At first I though I might just save the navigation state and reload it when a "device change" is detected, but it doesn't work because screens' "keys" do not match (different screen instances are used after a device switch).

I also tried saving the `window.location.pathname` and then navigate to that particular `pathname` (using `Linking.openURL`), but  it reloads the app and is not the right solution.

Is there a way to achieve this?

## Additional stuff that might be related

In case it matters, the useDeviceQuery is the following: 

```ts
import {useMediaQuery} from 'react-responsive';

import {MOBILE_MAX_WIDTH} from '~/constants/Layout';

export default function useDeviceQuery(): MediaQueryResponse {
    const isDesktop = useMediaQuery({query: `(min-width: ${MOBILE_MAX_WIDTH}px)`});
    return {isDesktop, isMobile: !isDesktop};
}

export type MediaQueryResponse = {
    isDesktop: boolean;
    isMobile: boolean;
};
```

**Comment by satya164** [comment]
I recommend building a custom navigator instead. The stack navigators are not built to handle different layouts.

https://reactnavigation.org/docs/custom-navigators/

If you think your case is a common use case, you can also create a feature request: https://react-navigation.canny.io/feature-requests
